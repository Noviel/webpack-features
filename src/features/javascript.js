import createBabelEnvPresetOptions from '../lib/babelEnvPresetOptions';

import { includeExclude } from '../lib/regexp';
import { tsConfigFile } from '../lib/tsConfig';

const syntaxExtendPlugins = [
  '@babel/plugin-proposal-object-rest-spread',
  '@babel/plugin-proposal-decorators',
  ['@babel/plugin-proposal-class-properties', { loose: true }],
  '@babel/plugin-syntax-dynamic-import',
];

const createPresetsList = ({ env, react, flow }) =>
  []
    .concat(
      env ? [['@babel/preset-env', createBabelEnvPresetOptions(env)]] : []
    )
    .concat(react ? '@babel/preset-react' : [])
    .concat(flow ? '@babel/preset-flow' : []);

export const createTestRegExp = exclude =>
  includeExclude(`\\.worker`, `\\.jsx?`, exclude);

export const createTestRegExpTS = exclude =>
  includeExclude(`\\.worker`, `\\.tsx?`, exclude);

export default (
  env,
  {
    eslint = true,
    flow = false,
    typescript = false,
    modules = false,
    react = false,
    webWorkers = true,
    syntaxExtend = true,
    hot = !env.production && env.target.name === 'browsers',
    babelPlugins = [],
    exclude = /node_modules/,
    tsOptions = {},
  } = {},
  { plugins = [], next }
) => {
  const { target } = env;

  const babelLoader = {
    loader: 'babel-loader',
    options: {
      babelrc: false,
      presets: createPresetsList({
        env: { target, modules },
        react,
        flow,
      }),
      plugins: []
        .concat(hot && react ? 'react-hot-loader/babel' : [])
        .concat(syntaxExtend ? syntaxExtendPlugins : [])
        .concat(babelPlugins)
        .concat(hot && react ? '@babel/transform-classes' : [])
        .concat(hot && react ? '@babel/transform-react-jsx-source' : []),
    },
  };

  const eslintLoader = {
    loader: 'eslint-loader',
    options: typeof eslint === 'object' ? eslint : {},
  };

  const tsLoader = {
    loader: 'ts-loader',
    options: {
      configFile: tsConfigFile(typescript),
      ...tsOptions,
    },
  };

  const migrationTS = typescript === 'migration';

  const rules = [
    {
      test: webWorkers ? createTestRegExp(true) : /\.jsx?/,
      exclude,
      use: []
        .concat(babelLoader)
        .concat(migrationTS ? tsLoader : [])
        .concat(eslint ? eslintLoader : []),
    },
  ];

  if (typescript) {
    rules.push({
      test: webWorkers ? createTestRegExpTS(true) : /\.tsx?/,
      exclude,
      // use babel-loader only for `migration` mode
      // in `strict` mode should be used pure TypeScript
      use: [].concat(migrationTS ? babelLoader : []).concat(tsLoader),
    });
  }

  if (webWorkers) {
    rules.unshift({
      test: createTestRegExp(false),
      exclude,
      use: {
        loader: 'worker-loader',
        options: {
          inline: !env.production,
          publicPath: env.publicPath,
          name: `[name].[hash].js`,
        },
      },
    });
  }

  const extensions = ['.js', '.jsx'].concat(typescript ? ['.ts', '.tsx'] : []);

  return next(env, plugins, {
    module: { rules },
    resolve: {
      extensions,
    },
  });
};
