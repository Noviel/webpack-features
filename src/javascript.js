import createBabelEnvPresetOptions from './babel-env-preset-options';

const syntaxExtendPlugins = [
  'transform-object-rest-spread',
  'transform-class-properties',
  'syntax-dynamic-import',
];

const createPresetsList = ({ env, react, flow }) =>
  []
    .concat(env ? [['env', createBabelEnvPresetOptions(env)]] : [])
    .concat(react ? 'react' : [])
    .concat(flow ? 'flow' : []);

const createEslintLoader = options => ({
  loader: 'eslint-loader',
  options: typeof options === 'object' ? options : {},
});

const createBabelLoader = (
  { target, production },
  { modules, react, flow, plugins, syntaxExtend, hot }
) => ({
  loader: 'babel-loader',
  options: {
    babelrc: false,
    presets: createPresetsList({ env: { ...target, modules }, react, flow }),
    plugins: []
      .concat(hot && react ? 'react-hot-loader/babel' : [])
      .concat(syntaxExtend ? syntaxExtendPlugins : [])
      .concat(plugins)
      .concat(hot && react ? 'transform-es2015-classes' : [])
      .concat(hot && react ? 'transform-react-jsx-source' : []),
  },
});

module.exports = (
  { target, production },
  {
    eslint = true,
    flow = true,
    modules = false,
    react = true,
    syntaxExtend = true,
    hot = !production && !!target.browsers,
    plugins = [],
  } = {},
  state
) => {
  state.addRules({
    test: /\.jsx?$/,
    exclude: /node_modules/,

    use: []
      .concat(
        createBabelLoader(
          {
            target,
            production,
          },
          { modules, react, flow, plugins, syntaxExtend, hot }
        )
      )
      .concat(eslint ? createEslintLoader(eslint) : []),
  });
};
