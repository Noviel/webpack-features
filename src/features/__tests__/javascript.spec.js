import javascript from '../javascript';
import { callFeature, envs, captionForEnv } from '../../fixtures';

describe('javascript', () => {
  for (const env in envs) {
    it(`should return a valid rule with default paremeters ${captionForEnv(
      envs[env]
    )}`, () => {
      expect(callFeature(javascript, {}, [], envs[env])).toMatchSnapshot();
    });
  }

  it('should apply migration TypeScript', () => {
    expect(
      callFeature(javascript, { typescript: 'migration' })
    ).toMatchSnapshot();
  });

  it('should apply strict TypeScript', () => {
    expect(callFeature(javascript, { typescript: 'strict' })).toMatchSnapshot();
  });
});
