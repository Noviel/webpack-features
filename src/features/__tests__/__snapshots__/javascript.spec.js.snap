// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`javascript should apply migration TypeScript 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.worker\\\\\\.jsx\\?\\$/i,
        "use": Object {
          "loader": "worker-loader",
          "options": Object {
            "inline": false,
            "name": "[name].[hash].js",
            "publicPath": "/",
          },
        },
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.jsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "browsers": Array [
                        "last 2 Chrome versions",
                        "not Chrome < 60",
                        "last 2 Safari versions",
                        "not Safari < 10.1",
                        "last 2 iOS versions",
                        "not iOS < 10.3",
                        "last 2 Firefox versions",
                        "not Firefox < 54",
                        "last 2 Edge versions",
                        "not Edge < 15",
                      ],
                    },
                    "useBuiltIns": false,
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "configFile": "C:\\\\projects\\\\dev\\\\webpack-features\\\\migration.tsconfig.json",
            },
          },
          Object {
            "loader": "eslint-loader",
            "options": Object {},
          },
        ],
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.tsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "browsers": Array [
                        "last 2 Chrome versions",
                        "not Chrome < 60",
                        "last 2 Safari versions",
                        "not Safari < 10.1",
                        "last 2 iOS versions",
                        "not iOS < 10.3",
                        "last 2 Firefox versions",
                        "not Firefox < 54",
                        "last 2 Edge versions",
                        "not Edge < 15",
                      ],
                    },
                    "useBuiltIns": false,
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "configFile": "C:\\\\projects\\\\dev\\\\webpack-features\\\\migration.tsconfig.json",
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
}
`;

exports[`javascript should apply strict TypeScript 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.worker\\\\\\.jsx\\?\\$/i,
        "use": Object {
          "loader": "worker-loader",
          "options": Object {
            "inline": false,
            "name": "[name].[hash].js",
            "publicPath": "/",
          },
        },
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.jsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "browsers": Array [
                        "last 2 Chrome versions",
                        "not Chrome < 60",
                        "last 2 Safari versions",
                        "not Safari < 10.1",
                        "last 2 iOS versions",
                        "not iOS < 10.3",
                        "last 2 Firefox versions",
                        "not Firefox < 54",
                        "last 2 Edge versions",
                        "not Edge < 15",
                      ],
                    },
                    "useBuiltIns": false,
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "eslint-loader",
            "options": Object {},
          },
        ],
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.tsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "ts-loader",
            "options": Object {
              "configFile": "C:\\\\projects\\\\dev\\\\webpack-features\\\\strict.tsconfig.json",
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
}
`;

exports[`javascript should return a valid rule with default paremeters for current node in production mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.worker\\\\\\.jsx\\?\\$/i,
        "use": Object {
          "loader": "worker-loader",
          "options": Object {
            "inline": false,
            "name": "[name].[hash].js",
            "publicPath": "/",
          },
        },
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.jsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "node": "current",
                    },
                    "useBuiltIns": false,
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "eslint-loader",
            "options": Object {},
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
    ],
  },
}
`;

exports[`javascript should return a valid rule with default paremeters for legacy browsers in development mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.worker\\\\\\.jsx\\?\\$/i,
        "use": Object {
          "loader": "worker-loader",
          "options": Object {
            "inline": true,
            "name": "[name].[hash].js",
            "publicPath": "/",
          },
        },
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.jsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "browsers": Array [
                        "> 1%",
                        "last 2 versions",
                        "Firefox ESR",
                      ],
                    },
                    "useBuiltIns": "entry",
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "eslint-loader",
            "options": Object {},
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
    ],
  },
}
`;

exports[`javascript should return a valid rule with default paremeters for legacy browsers in production mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.worker\\\\\\.jsx\\?\\$/i,
        "use": Object {
          "loader": "worker-loader",
          "options": Object {
            "inline": false,
            "name": "[name].[hash].js",
            "publicPath": "/",
          },
        },
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.jsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "browsers": Array [
                        "> 1%",
                        "last 2 versions",
                        "Firefox ESR",
                      ],
                    },
                    "useBuiltIns": "entry",
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "eslint-loader",
            "options": Object {},
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
    ],
  },
}
`;

exports[`javascript should return a valid rule with default paremeters for modern browsers in development mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.worker\\\\\\.jsx\\?\\$/i,
        "use": Object {
          "loader": "worker-loader",
          "options": Object {
            "inline": true,
            "name": "[name].[hash].js",
            "publicPath": "/",
          },
        },
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.jsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "browsers": Array [
                        "last 2 Chrome versions",
                        "not Chrome < 60",
                        "last 2 Safari versions",
                        "not Safari < 10.1",
                        "last 2 iOS versions",
                        "not iOS < 10.3",
                        "last 2 Firefox versions",
                        "not Firefox < 54",
                        "last 2 Edge versions",
                        "not Edge < 15",
                      ],
                    },
                    "useBuiltIns": false,
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "eslint-loader",
            "options": Object {},
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
    ],
  },
}
`;

exports[`javascript should return a valid rule with default paremeters for modern browsers in production mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.worker\\\\\\.jsx\\?\\$/i,
        "use": Object {
          "loader": "worker-loader",
          "options": Object {
            "inline": false,
            "name": "[name].[hash].js",
            "publicPath": "/",
          },
        },
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\^\\(\\?:\\(\\?!\\\\\\.worker\\)\\.\\)\\*\\\\\\.jsx\\?\\$/i,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "plugins": Array [
                "@babel/plugin-proposal-object-rest-spread",
                "@babel/plugin-proposal-decorators",
                Array [
                  "@babel/plugin-proposal-class-properties",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-syntax-dynamic-import",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "modules": false,
                    "targets": Object {
                      "browsers": Array [
                        "last 2 Chrome versions",
                        "not Chrome < 60",
                        "last 2 Safari versions",
                        "not Safari < 10.1",
                        "last 2 iOS versions",
                        "not iOS < 10.3",
                        "last 2 Firefox versions",
                        "not Firefox < 54",
                        "last 2 Edge versions",
                        "not Edge < 15",
                      ],
                    },
                    "useBuiltIns": false,
                  },
                ],
              ],
            },
          },
          Object {
            "loader": "eslint-loader",
            "options": Object {},
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
    ],
  },
}
`;
